swupd/packs.go

```diff
@@ -14,11 +14,306 @@
-	ToVersion     uint32
-	FullFileCount uint32
-	Manifest      *Manifest
+import (
+	"archive/tar"
+	"bytes"
+	"compress/bzip2"
+	"compress/gzip"
+	"fmt"
+	"io"
+	"log"
+	"os"
+	"path/filepath"
+)
+
+const debugPacks = false
+
+type PackState int
+
+const (
+	NotPacked PackState = iota
+	PackedDelta
+	PackedFullfile
+)
+
+type PackEntry struct {
+	File   *File
+	State  PackState
+	Reason string
+}
+
+type PackInfo struct {
+	// TODO: Add stats.
+	Entries  []PackEntry
+	Warnings []string
+}
+
+// WritePack writes the pack of a given Manifest from a version to the version of the Manifest. The
+// outputDir is used to pick deltas and fullfiles. If not empty, chrootDir is tried first as a fast
+// alternative to decompressing the fullfiles.
+func WritePack(w io.Writer, m *Manifest, fromVersion uint32, outputDir, chrootDir string) (info *PackInfo, err error) {
+	if m.Name == "" {
+		return nil, fmt.Errorf("manifest has no name")
+	}
+	if fromVersion >= m.Header.Version {
+		return nil, fmt.Errorf("fromVersion (%d) smaller than toVersion (%d)", fromVersion, m.Header.Version)
+	}
+
+	if debugPacks {
+		if chrootDir != "" {
+			log.Printf("DEBUG: using chrootDir=%s for packing", chrootDir)
+		} else {
+			log.Printf("DEBUG: not using chrootDir for packing")
+		}
+	}
+
+	info = &PackInfo{
+		Entries: make([]PackEntry, len(m.Files)),
+	}
+	defer func() {
+		if err != nil {
+			info = nil
+		}
+	}()
+
+	xw, err := newExternalWriter(w, "xz")
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		cerr := xw.Close()
+		if err == nil {
+			err = cerr
+		}
+	}()
+
+	tw := tar.NewWriter(xw)
+	err = tw.WriteHeader(&tar.Header{
+		Name:     "delta/",
+		Mode:     0700,
+		Typeflag: tar.TypeDir,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("write packed failed: %s", err)
+	}
+	err = tw.WriteHeader(&tar.Header{
+		Name:     "staged/",
+		Mode:     0700,
+		Typeflag: tar.TypeDir,
+	})
+	if err != nil {
+		return nil, fmt.Errorf("write packed failed: %s", err)
+	}
+
+	done := make(map[Hashval]bool)
+	for i, f := range m.Files {
+		entry := &info.Entries[i]
+		entry.File = f
+
+		if f.Version <= fromVersion {
+			entry.Reason = "already in from manifest"
+			continue
+		}
+		if done[f.Hash] {
+			entry.Reason = "hash already packed"
+			continue
+		}
+		if f.Status == statusDeleted {
+			entry.Reason = "file deleted"
+			continue
+		}
+		if f.Status == statusGhosted {
+			entry.Reason = "file ghosted"
+			continue
+		}
+
+		done[f.Hash] = true
+
+		// TODO: Pack deltas when available.
+		entry.State = PackedFullfile
+		entry.Reason = "from fullfile"
+		if chrootDir != "" {
+			err, fallback := copyFromChrootFile(tw, chrootDir, f)
+			if (err != nil) && fallback {
+				info.Warnings = append(info.Warnings, err.Error())
+				err = copyFromFullfile(tw, outputDir, f)
+			} else {
+				entry.Reason = "from chroot"
+			}
+		} else {
+			err = copyFromFullfile(tw, outputDir, f)
+		}
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	err = tw.Close()
+	if err != nil {
```

Comment:
you already deferred a func that will overwrite `info` with `nil` when `err != nil` so you should be able to just `return info, err` right here.

Reply:
That defer was useful when I had more cases of deferred functions causing error. Since there is only one now, moved the "info = nil" there.